Start and end of event, adding some features, beautifying, closing issues

KLVN opened the pull request:
Before:
![gcal_2](https://user-images.githubusercontent.com/8047980/34134607-e0a61878-e45b-11e7-872a-91316765ece4.PNG)

After:
![gcal_1](https://user-images.githubusercontent.com/8047980/34134126-16389018-e459-11e7-804e-679848d25e9d.PNG)



- Show start and end time of events
These are vertically aligned and centered. (See screenshot: "Long location")
Closes #307 #308 #356 

- Display multi-day events with given times as all-day events
It bothered me, that events, that span over multiple days and that are not all-day events (because they have start and end times) were shown as "normal" events that last only a few hours long. Now, these are also displayed as all-day events with their start and end times. (See screenshot: "Multiple days event with times")

- Go to already existing Calendar tab, instead of creating one every time
Not that good, as I wanted it to be. Are there any JS functions on Google Calendar to open an event, without reloading the entire website?
Closes #461 

- Beautifying
`<header>` is now outside the entire scrolling view, so the scrollbar is only used for the events, which looks better. I also redesigned the scrollbar itself and added some rounded borders for events and `quick-add`.

atdt commented on src/browser_action.css:
```diff
@@ -25,16 +50,28 @@ header {
+section {
+  -webkit-box-flex: 1;
+  -webkit-flex: 1;
+  -ms-flex: 1;
```
Isn't this for IE 10? This is a Chrome extension :)

atdt commented on src/browser_action.css:
```diff
@@ -25,16 +50,28 @@ header {
+}
+
+section {
+  -webkit-box-flex: 1;
```
Existing usage of 'flex' (e.g. for '.start-time' below) does not have any vendor-prefixed variants. We should be consistent. Do you know when Chrome started supporting flex?

KLVN replied:
> We should be consistent.

You are right, I added the other cases by habit.

From Chrome 21 to 28 with prefix `-webkit-` and since 29 with just using `flex`.

manastungare replied:
We should generally assume people are using Chrome evergreen (i.e. the latest version). The kinds of users who hunt down and install extensions like this usually keep their browser updated. Let’s drop the prefixes.

atdt commented on src/browser_action.css:
```diff
@@ -13,6 +13,31 @@ body {
   padding: 0;
   margin: 0;
 }
+::-webkit-scrollbar {
```
Some comment here about what you're doing to the scrollbar would be helpful.

KLVN replied:
I wasn't sure if I should include comments, because this `.css` file hardly has any.

manastungare replied:
Please feel free to add comments when they make sense, especially as more people start to contribute!

atdt commented:
Thank you for the pull-request! This looks good, for the most part.

I don't think tab re-use is a good idea, but @manastungare gave it a thumbs up in #461, so I'll wait for him to chime in. I'd also like him to approve the UI changes.

I left some comments in-line about some minor issues.

KLVN commented:
Thanks for your comments, regarding my PR. I will edit these things.

> I don't think tab re-use is a good idea, but @manastungare gave it a thumbs up in #461, so I'll wait for him to chime in.

I don't really like it either, but it was the only Issue labeled `Good First Issue`, so I wanted to give it a try :D Will delete these changes in my upcoming commits.


Should I include comments and if so, where should I put them and how detailed should they be? For example in the line itself, or above it; should I comment on every line I edited or are some things self-explaining and don't need a comment? I don't want to blow things up.
```js
if (!start) {  // Some events detected via microformats are malformed.
    return eventDiv;
}

// Some events detected via microformats are malformed.
if (!start) {
    return eventDiv;
}
```

KLVN commented on src/browser_action.js:
```diff
@@ -418,20 +418,19 @@ browseraction.createEventDiv_ = function(event) {
   var isHappeningNow = start.valueOf() < now && end.valueOf() >= now;
   var spansMultipleDays = (end.diff(start, 'seconds') > 86400);
-  if (event.allday) {
+  if (event.allday || !event.allday && spansMultipleDays) {
```
It is, just wanted to be more detailed by comparing all-day events with events spanning multiple days but that are not all-day events
... a bit overcomplicated :)

manastungare replied:
Worth extracting into a named boolean them?

```
var isMultiDayEventWithTime = …
if (event.allday || isMultiDayEventWithTime) { …
```

manastungare commented on src/browser_action.js:
```diff
@@ -489,7 +488,7 @@ browseraction.createEventDiv_ = function(event) {
   }
 
-  if (event.allday && spansMultipleDays || isDetectedEvent) {
+  if (event.allday && spansMultipleDays || !event.allday && spansMultipleDays || isDetectedEvent) {
```
Or just use a named condition to clarify things.

manastungare commented on src/browser_action.js:
```diff
@@ -498,6 +497,25 @@ browseraction.createEventDiv_ = function(event) {
+  return url.indexOf(getGCalUrl()) == 0;
+}
+
+function goToCalendar(eventUrl) {
```
Hmm, I think I prefer it this way. I often end up with multiple tabs I didn’t intend to open, which eventually annoys me because I start editing an event in one tab, and then switch away, and can’t find that tab any more after returning. More tabs also means Chrome requires more RAM.

To do the opposite, there’s already an easy way: simply open a new tab and start typing `calendar.google.com`, which is easy if that’s what you wanted.

So I’d prefer this change!

manastungare commented:
Thanks for this PR, and for joining the project!

manastungare commented:
In the future, please send one PR per bug fixed or feature implemented, instead of multiple changes in one. It makes it easy to review, easier to spot bugs because each PR is small and self-contained, and easy to revert a particular one if needed later (either because of a bug or feature change).

manastungare commented on src/browser_action.js:
```diff
@@ -498,6 +497,25 @@ browseraction.createEventDiv_ = function(event) {
 };
 
+function getGCalUrl() {
+  return "https://calendar.google.com/calendar/";
```
Which it already is: `constants.CALENDAR_UI_URL = 'https://calendar.google.com/calendar/';`

manastungare commented on src/browser_action.js:
```diff
@@ -497,6 +497,25 @@ browseraction.createEventDiv_ = function(event) {
+  return "https://calendar.google.com/calendar/";
+}
+
+function isGCalUrl(url) {
```
Inline this, after switching to use `constants.CALENDAR_UI_URL`

manastungare commented on src/browser_action.js:
```diff
@@ -497,6 +497,25 @@ browseraction.createEventDiv_ = function(event) {
+}
+
+function goToCalendar(eventUrl) {
+  chrome.tabs.getAllInWindow(undefined, function(tabs) {
```
Document the `undefined` parameter with an inline comment.

manastungare commented on src/browser_action.js:
```diff
@@ -497,6 +497,25 @@ browseraction.createEventDiv_ = function(event) {
+  chrome.tabs.getAllInWindow(undefined, function(tabs) {
+    for (var i = 0, tab; tab = tabs[i]; i++) {
+      if (tab.url && isGCalUrl(tab.url)) {
+        chrome.tabs.update(tab.id, {selected: true, url: eventUrl});
```
What happens if the user were already in the middle of editing an event in that tab?

I think we should only switch to an existing tab if the UI displayed in that tab is the Calendar UI Home Screen, which means the URL must match more rigorously.

KLVN commented:
Alright, so

`browser_action.css`
- deleted vendor prefixes
- deleted unnecessary code that I added by habit and for "just to be sure"-cases
- more width for times to support 12h format
- had two code blocks for <section> so I merged them together
- add comments

`browser_action.js`
- introduce `isMultiDayEventWithTime` for events spanning over multiple days and that have times (unlike all-day events)
- split ternary expression into `if...else` statements
- Re-use a calendar tab: as a compromise, it will only look for tabs in the current window. So splitting many events/different accounts into separate windows will work :) Also, a tab that is currently editing an event will not be "overwritten"
- add comments

manastungare commented on src/browser_action.css:
```diff
@@ -14,6 +14,30 @@ body {
+}
+
+::-webkit-scrollbar-thumb {
+  background: #afafaf; /* a bit darker as the original, because of the extensions background */
```
Darker _than_

manastungare commented on src/browser_action.js:
```diff
@@ -418,20 +418,28 @@ browseraction.createEventDiv_ = function(event) {
   var isHappeningNow = start.valueOf() < now && end.valueOf() >= now;
   var spansMultipleDays = (end.diff(start, 'seconds') > 86400);
-  if (event.allday) {
+  var isMultiDayEventWithTime = (!event.allday && spansMultipleDays); // Not an all-day event implies that times are given.
```
Please move the comment above, so the line width stays under 100 characters.

manastungare commented on src/browser_action.js:
```diff
@@ -418,20 +418,28 @@ browseraction.createEventDiv_ = function(event) {
-  var dateTimeFormat =
-      event.allday ? 'MMM D, YYYY' : (isDetectedEvent ? 'MMM D, YYYY ' + timeFormat : timeFormat);
+  
+  if(event.allday) { // Choose the correct time format.
```
Space after `if`. Need to run the formatter?

manastungare commented on src/browser_action.js:
```diff
@@ -498,6 +507,30 @@ browseraction.createEventDiv_ = function(event) {
+// Search for a Google Calendar tab and re-use this one, if none exists, then create a new one.
+function goToCalendar(eventUrl) {
+  chrome.tabs.query({
+    // All URLs showing Calendar UI Home Screen, except '/eventedit/', so current edits of events are not discarded.
```
Thanks, this is wonderful!

KLVN commented:
Done :)
